vulnerabilities/pipelines/v2_importers/project_kb_importer.py

```diff
@@ -0,0 +1,194 @@
+from pathlib import Path
+from typing import Iterable
+
+import saneyaml
+from fetchcode.vcs import fetch_via_vcs
+from packageurl import PackageURL
+from univers.version_range import RANGE_CLASS_BY_SCHEMES
+from univers.versions import InvalidVersion
+
+from vulnerabilities.importer import AdvisoryData
+from vulnerabilities.importer import AffectedPackageV2
+from vulnerabilities.pipelines import VulnerableCodeBaseImporterPipelineV2
+from vulnerabilities.pipes.advisory import append_patch_classifications
+from vulnerabilities.utils import get_advisory_url
+from vulnerabilities.utils import is_commit
+
+
+class ProjectKBPipeline(VulnerableCodeBaseImporterPipelineV2):
+    """
+    ProjectKB Importer Pipeline
+    Collect advisory from ProjectKB data:
+    - YAML statements: https://github.com/SAP/project-kb/blob/vulnerability-data/statements/*/*.yaml
+    - CSV database https://github.com/SAP/project-kb/blob/main/MSR2019/dataset/vulas_db_msr2019_release.csv
+    """
+
+    pipeline_id = "project-kb_v2"
+    spdx_license_expression = "Apache-2.0"
+    license_url = "https://github.com/SAP/project-kb/blob/main/LICENSE.txt"
+    main_branch = "git+https://github.com/SAP/project-kb"
+    vuln_data_branch = "git+https://github.com/SAP/project-kb@vulnerability-data"
+
+    @classmethod
+    def steps(cls):
+        return (cls.clone_repo, cls.collect_and_store_advisories, cls.clean_downloads)
+
+    def clone_repo(self):
+        self.log("Cloning ProjectKB advisory data...")
+        self.main_branch_vcs = fetch_via_vcs(self.main_branch)
+        self.vuln_data_branch_vcs = fetch_via_vcs(self.vuln_data_branch)
+
+    def advisories_count(self):
+        base_path = Path(self.vuln_data_branch_vcs.dest_dir) / "statements"
+        csv_path = (
+            Path(self.main_branch_vcs.dest_dir) / "MSR2019/dataset/vulas_db_msr2019_release.csv"
+        )
+
+        count_files = sum(1 for _ in base_path.rglob("*.yaml"))
+        with open(csv_path, newline="", encoding="utf-8") as f:
+            reader = csv.reader(f)
+            next(reader, None)
+            count_rows = sum(1 for _ in reader)
+
+        count = count_files + count_rows
+        self.log(f"Estimated advisories to process: {count}")
+        return count
+
+    def collect_advisories(self) -> Iterable[AdvisoryData]:
+        self.log("Collecting fix commits from YAML statements under /statements....")
+        base_path = Path(self.vuln_data_branch_vcs.dest_dir) / "statements"
+
+        for yaml_file in base_path.rglob("*.yaml"):
+            if yaml_file.name != "statement.yaml":
+                continue
+
+            with open(yaml_file, encoding="utf-8") as f:
+                yaml_data = saneyaml.load(f)
+
+            vulnerability_id = yaml_data.get("vulnerability_id")
+            if not vulnerability_id:
+                continue
+
+            note_texts = []
+            for note_entry in yaml_data.get("notes", []):
+                text_content = note_entry.get("text")
```

Comment:
Notes section may also contain list of links which should be collected and stored as references see example https://github.com/SAP/project-kb/blob/730d2192bfa9b909246e02bc051e4fad2958a6d9/statements/CVE-2018-16406/statement.yaml